### PR TITLE
Improve errors' UI

### PR DESCRIPTION
**Pull request by samwilson**
* Add new WsExportException that can be used to throw
  user-recoverable translatable errors.
* Fix various error-handling to use this new system.
* Move all HTTP-related errors out of the main export classes,
  and keep all that in the Controllers.
* Move CSS into its own file.

Bug: https://phabricator.wikimedia.org/T277237

### CODE REVIEW

File: src/Controller/ExportController.php [hunk 1]
```diff
@@ -99,8 +95,16 @@ public function home(
...
+		$exception = false;
+		$response = new Response();
 		if ( $request->get( 'page' ) ) {
-			return $this->export( $request, $api, $fontProvider, $generatorSelector, $creditRepo );
+			try {
+				return $this->export( $request, $api, $fontProvider, $generatorSelector, $creditRepo );
+			} catch ( WsExportException $ex ) {
+				$exception = $ex;
+				// @TODO Return different status codes for different errors (e.g. timeout vs not found).
+				$response->setStatusCode( 404 );
```
Comment (MusikAnimal) [inline src/Controller/ExportController.php:106]: If we're going to use a generic response code I would use a 400 or 500. A 404 should strictly be for not found, I think...

If it helps, maybe `WsExportException` could take an optional third argument that sets the status code that should returned here?

File: src/OpdsBuilder.php [hunk 2]
```diff
@@ -55,7 +60,7 @@ public function buildFromCategory( $categoryTitle ) {
 			'gcmlimit' => '100',
 		] );
 		if ( !array_key_exists( 'query', $response ) ) {
-			throw new NotFoundHttpException( "Category not found: $categoryTitle" );
+			throw new Exception( "Category not found: $categoryTitle" );
```
Comment (MusikAnimal) [inline src/OpdsBuilder.php:63]: Did we want this to be translatable, too? I think it's okay if we don't (probably a rare error), but as written the changes to this file don't have any user-facing effect, do they?

File: public/styles.css [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+.error > .message {
+	flex: 20;
+	padding-top: 0.4rem;
+}
+
+.error p {
+	margin: 0;
+}
```
Comment (MusikAnimal) [inline public/styles.css:39]: Nitpick, maybe also add
```
.error glyphicon {
    top: 9px;
}
```
to make the (X) icon in the error message more vertically aligned with the text. Tested in Chrome and Firefox.

**Review by MusikAnimal — COMMENTED** [review]
Looking good! Just a few minor concerns, mainly with using 404 as a generic response code. My feedback should be taken only as an opinion, though :)

File: src/WsExportException.php [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+	public function getI18nMessage(): string {
+		return 'exception-' . $this->i18nMessage;
+	}
+
+	public function getI18nParams(): array {
+		return $this->i18nParams;
+	}
+
+}
```
Comment (MusikAnimal) [inline src/WsExportException.php:29]: Should normally go in a directory called Exception. I know there's only one right now, but we might add more (I was thinking we need a wrapper for `ProcessFailedException` to keep the error message title consistent so the emails are grouped)

File: public/styles.css (hunk 3 above)
Comment (samwilson) [reply public/styles.css:89]: Good point. Actually, if we `align-items: center;` then it's always in the right place (I think).

File: src/Controller/ExportController.php (hunk 1 above)
Comment (samwilson) [reply src/Controller/ExportController.php]: Yeah I was thinking that would be good; was going to ask about it. Will change it now.

File: src/OpdsBuilder.php (hunk 2 above)
Comment (samwilson) [reply src/OpdsBuilder.php:63]: That's right: this is only used in a Command, and none of those are translated.

File: src/WsExportException.php (hunk 4 above)
Comment (samwilson) [reply src/WsExportException.php]: Done.

File: templates/bundles/TwigBundle/Exception/error.html.twig [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            </p>
+            <p>
+                <strong>{{ msg('error-page-details', [exception.message]) }}</strong>
+            </p>
+            <p>
+                {% set issue_link %}{#
+                    #}<a href="https://phabricator.wikimedia.org/maniphest/task/edit/form/1/?projects=ws-export">{#
+                        #}{{ msg( 'error-page-issue-link' ) }}{#
+                    #}</a>{#
+                #}{% endset %}
```
Comment (MusikAnimal) [inline templates/bundles/TwigBundle/Exception/error.html.twig:26]: Not sure if it works here but I think you can [remove the whitespace](https://symfony.com/blog/better-white-space-control-in-twig-templates) with:
```
{%- set issue_link %}
    <a href="https://phabricator.wikimedia.org/maniphest/task/edit/form/1/?projects=ws-export">
        {{- msg( 'error-page-issue-link' ) -}}
    </a>
{% endset -%}
```

Comment (samwilson) [reply templates/bundles/TwigBundle/Exception/error.html.twig:26]: Oh, that's a cool trick! Thanks. Done. It's only required at the end, to stop the period floating out on its own.

**Pull request merged**